File: ui/src/types/evaluation.ts
/**
 * TypeScript types for RAGAS evaluation results and runs.
 */

export interface MetricScore {
  score: number
  threshold: number
  passed: boolean
}

export type MetricName =
  | 'faithfulness'
  | 'answer_relevancy'
  | 'context_precision'
  | 'context_recall'
  | 'answer_correctness'

export interface RAGConfig {
  chunk_size?: number
  chunk_overlap?: number
  top_k?: number
  score_threshold?: number
  embedding_model?: string
}

export interface SLMConfig {
  model_id?: string
  quantization?: string
  temperature?: number
  max_new_tokens?: number
}

export interface ConfigSnapshot {
  rag?: RAGConfig
  slm?: SLMConfig
}

export interface PerQuestionResult {
  question: string
  answer: string
  ground_truth: string
  contexts: string[]
  source_document?: string
  scores?: Record<string, number>
  // flat fields for backwards compat
  faithfulness?: number
  answer_relevancy?: number
  context_precision?: number
  context_recall?: number
  answer_correctness?: number
}

export interface RAGASResult {
  run_id: string
  timestamp: string
  duration_s: number
  question_count: number
  status: 'passed' | 'failed'
  metrics: Record<MetricName, MetricScore>
  per_question: PerQuestionResult[]
  config_snapshot?: ConfigSnapshot
  fail_on_threshold_breach?: boolean
}

export interface EvalRun {
  run_id: string
  timestamp: string
  status: 'passed' | 'failed' | 'running' | 'queued' | 'failed'
  question_count: number
  duration_s: number
  metrics: Record<string, number>
}

export interface StartEvalRequest {
  dataset_path?: string
  metrics?: string[]
  run_id?: string
}

export interface StartEvalResponse {
  run_id: string
  message: string
}

// Metric display metadata
export const METRIC_LABELS: Record<string, string> = {
  faithfulness: 'Faithfulness',
  answer_relevancy: 'Answer Relevancy',
  context_precision: 'Context Precision',
  context_recall: 'Context Recall',
  answer_correctness: 'Answer Correctness',
}

export const METRIC_DESCRIPTIONS: Record<string, string> = {
  faithfulness: 'Is the answer grounded in the retrieved context? (no hallucination)',
  answer_relevancy: 'Does the answer address what the user actually asked?',
  context_precision: 'Are the retrieved chunks actually relevant to the question?',
  context_recall: 'Did retrieval surface all the chunks needed to answer correctly?',
  answer_correctness: 'Does the answer match the ground truth?',
}

File: backend/models/evaluation.py
"""
Pydantic models for RAGAS evaluation results.
"""

from __future__ import annotations

from typing import Any, Optional
from pydantic import BaseModel, Field


class MetricScore(BaseModel):
    """Score for a single RAGAS metric."""
    score: float = Field(..., ge=0.0, le=1.0, description="Metric score between 0 and 1.")
    threshold: float = Field(..., description="Minimum passing threshold.")
    passed: bool = Field(..., description="True if score >= threshold.")


class RAGConfig(BaseModel):
    """RAG configuration snapshot used during a run."""
    chunk_size: Optional[int] = None
    chunk_overlap: Optional[int] = None
    top_k: Optional[int] = None
    score_threshold: Optional[float] = None
    embedding_model: Optional[str] = None


class SLMConfig(BaseModel):
    """SLM configuration snapshot used during a run."""
    model_id: Optional[str] = None
    quantization: Optional[str] = None
    temperature: Optional[float] = None
    max_new_tokens: Optional[int] = None


class ConfigSnapshot(BaseModel):
    """Full config snapshot for a run."""
    rag: Optional[RAGConfig] = None
    slm: Optional[SLMConfig] = None


class PerQuestionResult(BaseModel):
    """Per-question score breakdown."""
    question: str
    answer: str
    ground_truth: str
    contexts: list[str] = Field(default_factory=list)
    source_document: Optional[str] = None
    faithfulness: Optional[float] = None
    answer_relevancy: Optional[float] = None
    context_precision: Optional[float] = None
    context_recall: Optional[float] = None
    answer_correctness: Optional[float] = None


class EvalRun(BaseModel):
    """Full result for a single RAGAS evaluation run."""
    run_id: str
    timestamp: str
    duration_s: float
    question_count: int
    status: str = Field(..., description="'passed' or 'failed'")
    metrics: dict[str, MetricScore]
    per_question: list[dict[str, Any]] = Field(default_factory=list)
    config_snapshot: Optional[ConfigSnapshot] = None
    fail_on_threshold_breach: bool = True


class EvalRunSummary(BaseModel):
    """Lightweight summary of an evaluation run for listing."""
    run_id: str
    timestamp: str
    status: str
    question_count: int
    duration_s: float
    metrics: dict[str, float] = Field(
        default_factory=dict,
        description="Flat map of metric_name -> score for quick display.",
    )
